0003-I-dapr-compose.md

```diff
@@ -0,0 +1,294 @@
+    command: ["python3" "app.py"]
+    restart: false
+  - app_id: backend
+    app_dir: ./backend/
+    app_protocol: GRPC
+    app_port: 3000
+    unix_domain_socket: "/tmp/test-socket"
+    env:
+      - DEBUG: false
+    command: ["./backend"]
+    restart: true
+```
+> Note: Running the dependencies for each app as contianers is out of scope for this discussion initially. We might consider that in the future.
+
+- Each file conatins a `common` object which contains `env`, `resources_dir` and `config_file` that can be used in common across all the apps defined in this YAML
+- There is an `apps` section that lists the different app configs.
+- Each app config has the following
+  - `app_id` application ID (mandatory field). Passed to `daprd` as `--app-id`.
+  - `app_dir` directory of the application (mandatory field).
+  - `resources_dir` (optional) directory(ies) of all dapr resources (components, resiliency policies, subscription crds) (overrides common def). Passed to `daprd` as `--resources-dir`.
+  - `config_file` (optional) the configuration file to be used for this app (overrides common def). Passed to `daprd` as `--config-file`.
+  - `app_protocol` Application protocol, HTTP, gRPC defaults to HTTP. Passed to `daprd` as `--app-protocol`.
+  - `app_port` port the app listens to if any. Passed to `daprd` as `--app-port`.
+  - other dapr run parameters (mostly pass through flags to `daprd`) All properties must have `_` as separators which will be validated(so that no unknown flags are passed) and translated to `-` for cmd line arguments for `daprd`.
+  - `command` ["exec" "arg1" "arg2"] format for application command
+  - `env` which overrides or adds to common env var defined or the shell env var passed in when `dapr compose` is called
+  - `restart` field which lets `dapr compose` know if the app/sidecar should be restarted on crash or not
+
+The DAPR_HTTP_PORT and DAPR_GRPC_PORT will be passed in as extra environment variables to the application that is being run. Those flags for `daprd` can be overridden in the config file above but that is optional as the `compose` command will randomly assign a free port for each of those ports, similar to how `run` command currently behaves.
+
+### Preedence rules
+
+For `env` field:
+> Note: In addition to the defined env fields the app also gets the `DAPR_HTTP_PORT` and `DAPR_GRPC_PORT` fields.
+
+- If no field is present, the environment variables of the current shell which executes the `compose` command is passed to the app and dir.
+- if `env` field is present in the `common` section, in addition to the shell environment variables, the `env` map defined will be passed to all `apps` and `daprd` sidecars.
+- if `env` field is present only in a particular `apps` section, any shell environment variables, `env` maps from `common` section and the `env` map for the current app will be passed to both the `app` and `daprd`.
```

Comment:
Since the different apps will be triggered from the the same shell that the daemon is running, the env variables might conflict?